feat: Add support for multiple options in Select

tqwewe opened the pull request:
Adds support for selecting multiple options in the `Select` component using closable tags.

This should close https://github.com/thaw-ui/thaw/issues/101

https://github.com/thaw-ui/thaw/assets/16362377/a9e98c60-31a9-4c7c-8281-ec6dd6548b67

**Issues**:
- ~~Since this changes the type of `value` in the `Select` component, it means this PR is a breaking change.~~
- ~~The popup menu doesn't update it's position when the multiselect gets too long and goes onto two lines.~~ (Fixed with https://github.com/thaw-ui/thaw/pull/166/commits/d0e96ebe7f8318861d4177134ac6be6c7dffbe60)

tqwewe commented:
I was considering that, it would also mean it's no longer a breaking change. Though I'm not sure how this could be done without duplicating so much of the code?

luoxiaozero commented:
like  https://www.naiveui.com/en-US/os-theme/components/select#max-tag-count.vue ?

tqwewe commented:
Hmm not really limiting, I mean more like custom render function for the label itself.
https://ant.design/components/select#select-demo-custom-label-render

For users who want to render anything they want. In my personal use case, I want a filter called "Bookmakers", and I don't want to render all the bookmakers selected, instead I only want to show the number of bookmakers selected.

```
[ Bookmakers (3) ]
- [x] Betfair
- [x] Sportsbet
- [x] Tab
- [ ] Pinnacle
```
This could be done with something like:
```rust
view! {
    <MultiSelect
        label=|values| view! { {format!("Bookmakers ({})", values.len()) }
        values
        options
    />
}
```

luoxiaozero commented:
We can do this by adding a label slot.

```rust
view! {
    <Multiselect
        values
        options
    >
         <MultiselectLabel slot>
              {format!("Bookmakers ({})", values.len())
        <MultiselectLabel>
   </Multiselect>
}
```

tqwewe commented:
@luoxiaozero I've separated them into separate components. I also renamed it from `MultiSelect` into `SelectMulti` so it would be found in the docs beside `Select` (alphabetical).

I also implemented the `SelectLabel` slot for the selects.

We should also probably include a dropdown arrow in the select component. But maybe that can be a separate PR?

luoxiaozero commented on thaw/src/select/mod.rs:
```diff
@@ -1,177 +1,89 @@
 pub struct SelectOption<T> {
     pub label: String,
     pub value: T,
+    pub variant: TagVariant,
```
There will be destructive updates here, so maybe we can extract it as MultiSelectOption better.

luoxiaozero commented on thaw/src/select_multi/mod.rs:
It is best to move it to the Select directory and changed the name to MultiSelect.

tqwewe replied:
This is what I had done initially, but I thought to keep it consistent with the other components I would separate it because I noticed `Input`/`InputNumber` are completely separate in docs and folders. But I will merge move the multiselect to the select directory.

luoxiaozero replied:
I'm a little bit conventional, because other components do this too. Keep it that way.

luoxiaozero commented on thaw/src/select_multi/mod.rs:
```diff
@@ -0,0 +1,119 @@
+
+#[component]
+pub fn SelectMulti<T>(
+    #[prop(optional, into)] values: Model<Vec<T>>,
```
For consistency of field names, it is better to change to value.

tqwewe commented on demo_markdown/docs/select_multi/mod.md:
If so, how should the documentation look like? Should the props be listed separately?
Eg:
```md
### Select Props

| Name    | Type                                | Default              | Description                               |
| ------- | ----------------------------------- | -------------------- | ----------------------------------------- |
| class   | `OptionalProp<MaybeSignal<String>>` | `Default::default()` | Addtional classes for the select element. |
| value   | `Model<Option<T>>`                  | `None`               | Checked value.                            |
| options | `MaybeSignal<Vec<SelectOption<T>>>` | `vec![]`             | Options that can be selected.             |

### Multiple Select Props

| Name    | Type                                | Default              | Description                               |
| ------- | ----------------------------------- | -------------------- | ----------------------------------------- |
| class   | `OptionalProp<MaybeSignal<String>>` | `Default::default()` | Addtional classes for the select element. |
| values  | `Model<Vec<T>>`                     | `vec![]`             | Checked values.                           |
| options | `MaybeSignal<Vec<SelectOption<T>>>` | `vec![]`             | Options that can be selected.             |

### Select/Multiple Select Slots

| Name        | Default | Description   |
| ----------- | ------- | ------------- |
| SelectLabel | `None`  | Select label. |
```

luoxiaozero replied:
Keep it that way.

tqwewe commented:
Great I've pushed another commit with these changes

tqwewe commented:
Actually I've just gone ahead and added the dropdown arrow to the select component. Was quite a simple task so thought I'd add it here.
<img width="475" alt="Screenshot 2024-04-20 at 2 03 05 PM" src="https://github.com/thaw-ui/thaw/assets/16362377/ca20ab8a-3004-48c1-a75d-38c917fdc47a">

tqwewe commented:
I'm just trying to debug why it's looking a little different in my personal project when using this fork. The alignment isn't correct.
<img width="406" alt="Screenshot 2024-04-20 at 2 22 53 PM" src="https://github.com/thaw-ui/thaw/assets/16362377/94d5a67e-78b1-4fc1-af42-47e14494082d">

I've been digging through the styles but not sure what's different.

tqwewe commented on thaw_components/src/binder/mod.rs:
```diff
@@ -182,7 +184,8 @@ fn FollowerContainer<El: ElementDescriptor + Clone + 'static>(
             let width = match width {
-                FollowerWidth::Target => format!("width: {}px;", target_rect.width()),
+                FollowerWidth::Target => format!("min-width: {}px;", target_rect.width()),
+                FollowerWidth::MinTarget => format!("width: {}px;", target_rect.width()),
```
Oops, `MinTarget` should've been `min-width` instead of `Target`.
I added `MinTarget` and used it in the `Select` now to make sure the select dropdow menu doesn't get cut off if the select input is too short. I think this would be fine to have the dropdown menu be longer in case the options are longer, but its up to you to keep this or not. Personally as a user of Thaw, I'd prefer the text in the dropdown not to be cut off

tqwewe commented on thaw/src/select/select.css:
```diff
@@ -1,9 +1,11 @@
-    padding: 0 10px;
-    height: 30px;
-    line-height: 30px;
+    display: inline-block;
```
Okay I can remove this in order to keep the PR non-breaking, but are you happy to include this in the next version? If so, I can open an issue to track this.

I often refer to Ant Design and other UI libraries for these decisions and see they use inline-block, and it also makes sense for the UI's I'm building with Thaw.

luoxiaozero replied:
I need to think about it. I opened #176 .

tqwewe commented:
I'd also like to share this screenshot of what I've implemented in my project where I'm using this fork and the `<SelectLabel slot>` slot to customize the label.

<img width="471" alt="Screenshot 2024-04-22 at 10 53 35 PM" src="https://github.com/thaw-ui/thaw/assets/16362377/1deee161-85be-4cf5-a792-a5f50ef23106">

It's an alternative version of the multi select, where it has a label, and shows a `<Tag>` beside it with the number of items selected. Perhaps something like this might be of interest to people in the future?

luoxiaozero commented:
No more problems, now merge.
